onnxruntime/core/xnnpack/schema/xnnpack_onnx_defs.cc

```diff
@@ -0,0 +1,395 @@
+using ::ONNX_NAMESPACE::Common::StatusCode;
+
+OnnxStatus ComputeOutputSizeSame(ptrdiff_t input_size, uint32_t stride, ptrdiff_t* output_size) {
+  if (stride == 0 || input_size <= 0) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  size_t r = static_cast<size_t>(input_size) + stride - 1;
+  if (r > static_cast<size_t>(std::numeric_limits<ptrdiff_t>::max())) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  *output_size = r / stride;
+  return OnnxStatus::OK();
+}
+
+OnnxStatus ComputeOutputSizeValid(ptrdiff_t input_size, uint32_t stride, ptrdiff_t filter_size, uint32_t dilation_rate,
+                                  ptrdiff_t* output_size) {
+  if (filter_size < 1) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  if (stride <= 0) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  if (dilation_rate > 1) {
+    if (!SafeMultiply(filter_size - 1, dilation_rate, filter_size)) {
+      return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+    }
+    if (!SafeAdd(filter_size, 1, filter_size)) {
+      return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+    }
+  }
+
+  if (!SafeSubtract(input_size, filter_size, input_size)) return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  if (input_size < 0) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  if (!SafeAdd(input_size, stride, input_size)) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  assert(input_size > 0);
+  if (!SafeDivide(input_size, stride, input_size)) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL);
+  }
+  *output_size = input_size;
+  return ::ONNX_NAMESPACE::Common::Status::OK();
+}
+
+OnnxStatus XnnPackConvShapeInferImpl(const ::ONNX_NAMESPACE::TensorShapeProto& input_shape,
+                                     const ::ONNX_NAMESPACE::TensorShapeProto& weight_shape, uint32_t input_padding_top,
+                                     uint32_t input_padding_right, uint32_t input_padding_bottom,
+                                     uint32_t input_padding_left, uint32_t subsampling_height,
+                                     uint32_t subsampling_width, uint32_t dilation_h, uint32_t dilation_w,
+                                     ONNXRUNTIME_XNNPACK_PADDING_MODE padding_mode,
+                                     ::ONNX_NAMESPACE::TensorShapeProto* final_output_shape) {
+  if (input_shape.dim_size() != 4) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Input tensor must have 4 dimensions");
+  }
+
+  if (weight_shape.dim_size() != 4) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Weight tensor must have 4 dimensions");
+  }
+
+  if (!input_shape.dim(3).has_dim_value()) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Channel can not be unknown");
+  }
+  // Though the first dim of weight can be unknown, our implementation doesn't support it.
```

Comment:
If implementation doesn't support unknown dim 0, why not fail fast here?

Reply:
are you handling this in the transformer?

Reply:
This shape inference function doesn't require it. But the kernel implementation expects it. So, I can make the check stricter here. It doesn't change much. The transformer expects weights being constant.